domain/src/main/java/org/oppia/domain/exploration/ExplorationDataController.kt

```diff
@@ -0,0 +1,289 @@
+package org.oppia.domain.exploration
+
+import android.content.Context
+import androidx.lifecycle.LiveData
+import org.json.JSONArray
+import org.json.JSONException
+import javax.inject.Inject
+import javax.inject.Singleton
+import org.oppia.app.model.Exploration
+import org.oppia.util.data.AsyncResult
+import java.io.IOException
+import org.json.JSONObject
+import org.oppia.app.model.AnswerGroup
+import org.oppia.app.model.Outcome
+import org.oppia.app.model.Interaction
+import org.oppia.app.model.InteractionObject
+import org.oppia.app.model.RuleSpec
+import org.oppia.app.model.State
+import org.oppia.app.model.StringList
+import org.oppia.app.model.SubtitledHtml
+import org.oppia.util.data.DataProviders
+
+const val TEST_EXPLORATION_ID_0 = "test_exp_id_0"
+const val TEST_EXPLORATION_ID_1 = "test_exp_id_1"
+private const val WELCOME_EXPLORATION_DATA_PROVIDER_ID = "WelcomeExplorationDataProvider"
+private const val ABBOUT_OPPIA_EXPLORATION_DATA_PROVIDER_ID = "AboutOppiaExplorationDataProvider"
+
+/** Controller for retrieving an exploration. */
+@Singleton
+class ExplorationDataController @Inject constructor(private val context: Context,
+                                                    private val dataProviders: DataProviders
+) {
+
+  private val welcomeExplorationDataProvider =
```

Comment:
I suggest using a single provider for all explorations and selecting the correct data to load based on the ID.

To simplify things, I don't think we need to persist the providers as fields since the exploration structure should not change during the lifetime of the observer. This means we can create a new local provider for each call to ``getExplorationId``.

Reply:
Done